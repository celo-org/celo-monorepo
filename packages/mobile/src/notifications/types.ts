export enum NotificationTypes {
  PAYMENT_RECEIVED = 'PAYMENT_RECEIVED',
  PAYMENT_REQUESTED = 'PAYMENT_REQUESTED',
}

export interface TransferNotificationData {
  recipient: string
  sender: string
  value: string
  blockNumber: string
  txHash: string
  timestamp: string
  comment: string
  currency: string
  type?: NotificationTypes.PAYMENT_RECEIVED
}

export enum NotificationReceiveState {
  APP_ALREADY_OPEN,
  APP_FOREGROUNDED,
  APP_OPENED_FRESH,
}
