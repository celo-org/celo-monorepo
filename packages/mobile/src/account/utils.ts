import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
