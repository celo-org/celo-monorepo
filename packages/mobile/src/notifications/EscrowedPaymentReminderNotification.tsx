import BaseNotification from '@celo/react-components/components/BaseNotification'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Image, Linking, Platform, StyleSheet, View } from 'react-native'
import SendIntentAndroid from 'react-native-send-intent'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { componentWithAnalytics } from 'src/analytics/wrapper'
import { EscrowedPayment } from 'src/escrow/actions'
import EscrowedPaymentLineItem from 'src/escrow/EscrowedPaymentLineItem'
import { Namespaces } from 'src/i18n'
import { inviteFriendsIcon } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import Logger from 'src/utils/Logger'

interface OwnProps {
  payment: EscrowedPayment
}

type Props = OwnProps & WithNamespaces

export class EscrowedPaymentReminderNotification extends React.PureComponent<Props> {
  getCTA = () => {
    const { payment, t } = this.props
    const recipientPhoneNumber = payment.recipientPhone
    return [
      {
        text: t('sendMessage'),
        onPress: () => {
          CeloAnalytics.track(CustomEventNames.clicked_escrowed_payment_send_message)
          // TODO: open up whatsapp/text message slider with pre populated message
          try {
            if (Platform.OS === 'android') {
              SendIntentAndroid.sendSms(recipientPhoneNumber, t('escrowedPaymentReminderSms'))
            } else {
              // TODO look into using MFMessageComposeViewController to prefill the body for iOS
              Linking.openURL(`sms:${recipientPhoneNumber}`)
            }
          } catch {
            Logger.showError(t('SMSError'))
            Logger.error(
              'EscrowedPaymentReminderNotification/',
              t('SMSErrorDetails', {
                recipientNumber: recipientPhoneNumber,
              })
            )
          }
        },
      },
      {
        text: this.props.t('reclaimPayment'),
        onPress: () => {
          const reclaimPaymentInput = payment
          CeloAnalytics.track(CustomEventNames.clicked_escrowed_payment_notification)
          navigate(Screens.ReclaimPaymentConfirmationScreen, { reclaimPaymentInput })
        },
      },
    ]
  }

  getTitle() {
    const { t, payment } = this.props
    const displayName = payment.recipientContact
      ? payment.recipientContact.displayName
      : payment.recipientPhone
    return t('escrowedPaymentReminder', { mobile: displayName })
  }

  render() {
    const { payment } = this.props

    return (
      <BaseNotification
        title={this.getTitle()}
        icon={<Image source={inviteFriendsIcon} style={styles.image} resizeMode="contain" />}
        ctas={this.getCTA()}
        roundedBorders={true}
      >
        <View style={styles.body}>
          <View style={styles.payment}>
            <EscrowedPaymentLineItem payment={payment} />
          </View>
        </View>
      </BaseNotification>
    )
  }
}

const styles = StyleSheet.create({
  body: {
    marginTop: 5,
    flexDirection: 'row',
  },
  image: {
    width: 30,
    height: 30,
  },
  payment: {
    flex: 1,
  },
})

export default componentWithAnalytics(
  withNamespaces(Namespaces.walletFlow5)(EscrowedPaymentReminderNotification)
)
