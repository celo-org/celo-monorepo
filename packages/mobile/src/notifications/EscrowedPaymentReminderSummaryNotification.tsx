import BaseNotification from '@celo/react-components/components/BaseNotification'
import variables from '@celo/react-components/styles/variables'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Image, StyleSheet, View } from 'react-native'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { EscrowedPayment } from 'src/escrow/actions'
import EscrowedPaymentLineItem from 'src/escrow/EscrowedPaymentLineItem'
import { Namespaces } from 'src/i18n'
import { inviteFriendsIcon } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Stacks } from 'src/navigator/Screens'

interface OwnProps {
  payments: EscrowedPayment[]
}

type Props = OwnProps & WithNamespaces

const PREVIEW_SIZE = 2

export class EscrowedPaymentReminderSummaryNotification extends React.Component<Props> {
  getTotal() {
    return this.props.payments.length
  }

  getTitle() {
    const { t } = this.props
    return this.getTotal() > 1
      ? t('escrowedPaymentReminderWithCount_plural', { count: this.getTotal() })
      : t('escrowedPaymentReminder')
  }
  getCTA = () => {
    return [
      {
        text: this.props.t('review'),
        onPress: () => {
          CeloAnalytics.track(CustomEventNames.escrowed_payment_review)
          navigate(Stacks.EscrowStack)
        },
      },
    ]
  }
  render() {
    const { payments } = this.props
    return (
      <BaseNotification
        title={this.getTitle()}
        icon={<Image source={inviteFriendsIcon} style={styles.image} resizeMode="contain" />}
        ctas={this.getCTA()}
        roundedBorders={true}
      >
        <View style={styles.body}>
          <View style={styles.requests}>
            {payments.slice(0, PREVIEW_SIZE).map((payment, key) => {
              return <EscrowedPaymentLineItem payment={payment} key={key} />
            })}
          </View>
        </View>
      </BaseNotification>
    )
  }
}

const styles = StyleSheet.create({
  body: {
    marginTop: 5,
    flexDirection: 'row',
  },
  image: {
    width: 30,
    height: 30,
  },
  requests: {
    flex: 1,
  },
  counter: {
    paddingLeft: variables.contentPadding,
    justifyContent: 'flex-end',
  },
})

export default withNamespaces(Namespaces.walletFlow5)(EscrowedPaymentReminderSummaryNotification)
