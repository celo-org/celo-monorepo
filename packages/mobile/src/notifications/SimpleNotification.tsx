import BaseNotification, { CTA } from '@celo/react-components/components/BaseNotification'
import fontStyles from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { Image, ImageSourcePropType, StyleSheet, Text, View } from 'react-native'

interface OwnProps {
  text: string
  title: string
  image: ImageSourcePropType
  ctaList: CTA[]
}

type Props = OwnProps

export function SimpleNotification(props: Props) {
  return (
    <BaseNotification
      title={props.title}
      icon={<Image source={props.image} style={styles.image} resizeMode="contain" />}
      ctas={props.ctaList}
    >
      <View style={styles.body}>
        <Text style={fontStyles.subSmall}>{props.text}</Text>
      </View>
    </BaseNotification>
  )
}

const styles = StyleSheet.create({
  body: {
    marginTop: 5,
  },
  image: {
    width: 30,
    height: 30,
  },
})

export default SimpleNotification
