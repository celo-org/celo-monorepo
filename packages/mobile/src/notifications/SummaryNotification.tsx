import SummaryNotification from '@celo/react-components/components/SummaryNotification'
import fontStyles from '@celo/react-components/styles/fonts'
import * as React from 'react'
import i18n from 'src/i18n'

import { StyleSheet, Text, View } from 'react-native'

interface OwnProps<T> {
  title: string
  icon: JSX.Element
  onReview: () => void
  itemRenderer: (item: T, key: number) => JSX.Element
  items: T[]
}

type Props<T> = OwnProps<T>

const PREVIEW_SIZE = 2

function getAdditionalItemsCount<T>(items: T[]) {
  const total = items.length
  if (total - PREVIEW_SIZE > 0) {
    return (
      <View>
        <Text style={styles.moreWithCountText}>
          {i18n.t('global:moreWithCount', { count: total - PREVIEW_SIZE + 1 })}
        </Text>
      </View>
    )
  }
}

// Payment Request notification for the notification center on home screen
function PaymentsSummaryNotification<T>(props: Props<T>) {
  const { items, title, icon, onReview, itemRenderer } = props

  return (
    <SummaryNotification
      title={title}
      icon={icon}
      reviewCTA={{
        text: i18n.t('walletFlow5:review'),
        onPress: onReview,
      }}
    >
      <View style={styles.body}>
        <View style={styles.items}>
          {items
            .slice(0, props.items.length > PREVIEW_SIZE ? PREVIEW_SIZE - 1 : PREVIEW_SIZE)
            .map(itemRenderer)}
          {getAdditionalItemsCount(items)}
        </View>
      </View>
    </SummaryNotification>
  )
}

const styles = StyleSheet.create({
  body: {
    marginTop: 5,
    flexDirection: 'row',
  },
  items: {
    flex: 1,
  },
  moreWithCountText: fontStyles.subSmall,
})

export default PaymentsSummaryNotification
