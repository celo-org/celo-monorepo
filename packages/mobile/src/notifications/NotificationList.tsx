import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import React, { useEffect } from 'react'
import { ScrollView, StyleSheet, Text, View } from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import {
  NavigationParams,
  NavigationProp,
  NavigationScreenProp,
  NavigationState,
} from 'react-navigation'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import i18n from 'src/i18n'
import { headerWithBackButton } from 'src/navigator/Headers'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import { getCentAwareMoneyDisplay } from 'src/utils/formatting'

const { contentPadding } = variables

interface OwnProps<T> {
  dollarBalance: string | null
  listItemRenderer: (item: T, key: number) => JSX.Element
  items: T[]
}

type Props<T> = OwnProps<T>

export function NotificationList<T>(props: Props<T>) {
  return (
    <SafeAreaView style={styles.container}>
      <DisconnectBanner />
      {props.items.length > 0 ? (
        <ScrollView>
          <View style={styles.scrollArea}>{props.items.map(props.listItemRenderer)}</View>
        </ScrollView>
      ) : (
        <Text style={[fontStyles.bodySecondary, styles.empty]}>{i18n.t('global:emptyList')}</Text>
      )}
    </SafeAreaView>
  )
}

export function titleWithBalanceNavigationOptions(title: string) {
  return ({ navigation }: { navigation: NavigationProp<NavigationState> }) => ({
    ...headerWithBackButton,
    headerTitle: (
      <View style={styles.header}>
        {title && <Text style={fontStyles.bodyBold}>{title}</Text>}
        <Text style={styles.balanceText}>
          {(navigation.state.params &&
            navigation.state.params.dollarBalance &&
            CURRENCIES[CURRENCY_ENUM.DOLLAR].symbol +
              i18n.t('sendFlow7:celoDollarsAvailable', {
                CeloDollars: getCentAwareMoneyDisplay(navigation.state.params.dollarBalance || 0),
              })) ||
            i18n.t('global:loading')}
        </Text>
      </View>
    ),
  })
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.background,
    flex: 1,
  },
  scrollArea: {
    margin: contentPadding,
  },
  balanceText: {
    ...fontStyles.bodySmall,
    color: colors.inactiveDark,
  },
  empty: {
    textAlign: 'center',
    marginTop: 30,
  },
  header: {
    alignItems: 'center',
    flex: 1,
  },
})

// Should be used together with titleWithBalanceNavigationOptions.
// Call this hook inside a Screen and assign
// Screen.navigationOptions = titleWithBalanceNavigationOptions(title)
export function useBalanceInNavigationParam(
  dollarBalance: string | null,
  navigation: NavigationScreenProp<NavigationParams>
) {
  useEffect(
    () => {
      navigation.setParams({ dollarBalance })
    },
    [dollarBalance]
  )
}
