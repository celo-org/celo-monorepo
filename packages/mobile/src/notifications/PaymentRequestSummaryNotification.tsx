import BaseNotification from '@celo/react-components/components/BaseNotification'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Image, StyleSheet, Text, View } from 'react-native'
import { connect } from 'react-redux'
import { PaymentRequest } from 'src/account'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { Namespaces } from 'src/i18n'
import {
  addressToE164NumberSelector,
  AddressToE164NumberType,
  e164NumberToAddressSelector,
  E164NumberToAddressType,
} from 'src/identity/reducer'
import { sendDollar } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Stacks } from 'src/navigator/Screens'
import PaymentRequestNotificationInner from 'src/paymentRequest/PaymentRequestNotificationInner'
import { NumberToRecipient, phoneNumberToRecipient } from 'src/recipients/recipient'
import { recipientCacheSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'

interface OwnProps {
  requests: PaymentRequest[]
}

type Props = OwnProps & WithNamespaces & StateProps

interface StateProps {
  e164PhoneNumberAddressMapping: E164NumberToAddressType
  addressToE164Number: AddressToE164NumberType
  recipientCache: NumberToRecipient
}

const mapStateToProps = (state: RootState): StateProps => ({
  e164PhoneNumberAddressMapping: e164NumberToAddressSelector(state),
  addressToE164Number: addressToE164NumberSelector(state),
  recipientCache: recipientCacheSelector(state),
})

const PREVIEW_SIZE = 2

// Payment Request notification for the notification center on home screen
export class PaymentRequestSummaryNotification extends React.Component<Props> {
  getCTA = () => {
    return [
      {
        text: this.props.t('review'),
        onPress: () => {
          CeloAnalytics.track(CustomEventNames.request_payment_review)
          navigate(Stacks.RequestStack)
        },
      },
    ]
  }

  getTotal() {
    return this.props.requests.length
  }

  getAdditionalRequestsCount() {
    if (this.getTotal() - PREVIEW_SIZE > 0) {
      return (
        <View style={styles.counter}>
          <Text style={styles.requestCountText}>+{this.getTotal() - PREVIEW_SIZE}</Text>
        </View>
      )
    }
  }

  getTitle() {
    const { t } = this.props
    return this.getTotal() > 1
      ? t('paymentRequestWithCount', { count: this.getTotal() })
      : t('paymentRequest')
  }

  getRequesterRecipient = (requesterE164Number: string) => {
    return phoneNumberToRecipient(
      requesterE164Number,
      this.props.e164PhoneNumberAddressMapping[requesterE164Number],
      this.props.recipientCache
    )
  }

  render() {
    const { requests } = this.props

    return (
      <BaseNotification
        title={this.getTitle()}
        icon={<Image source={sendDollar} style={styles.image} resizeMode="contain" />}
        ctas={this.getCTA()}
        roundedBorders={true}
      >
        <View style={styles.body}>
          <View style={styles.requests}>
            {requests.slice(0, PREVIEW_SIZE).map((item, key) => {
              return (
                <PaymentRequestNotificationInner
                  key={key}
                  amount={item.amount}
                  comment={item.comment}
                  requesterE164Number={item.requesterE164Number}
                  requesterRecipient={this.getRequesterRecipient(item.requesterE164Number)}
                />
              )
            })}
          </View>
          {this.getAdditionalRequestsCount()}
        </View>
      </BaseNotification>
    )
  }
}

const styles = StyleSheet.create({
  body: {
    marginTop: 5,
    flexDirection: 'row',
  },
  image: {
    width: 30,
    height: 30,
  },
  requests: {
    flex: 1,
  },
  counter: {
    paddingLeft: variables.contentPadding,
    justifyContent: 'flex-end',
  },
  requestCountText: {
    ...fontStyles.semiBold,
    fontSize: 13,
    lineHeight: 17,
    color: colors.celoGreen,
  },
})

export default connect<StateProps, {}, {}, RootState>(mapStateToProps)(
  withNamespaces(Namespaces.walletFlow5)(PaymentRequestSummaryNotification)
)
