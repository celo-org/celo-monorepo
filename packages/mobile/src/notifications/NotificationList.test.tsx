import * as React from 'react'
import { Text } from 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import { NotificationList } from 'src/notifications/NotificationList'
import { createMockStore } from 'test/utils'

const props = () => ({
  dollarBalance: '123',
  items: ['a'],
  listItemRenderer: (item: string, key: number) => {
    return <Text>{`test-${item}`}</Text>
  },
})

describe(NotificationList, () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={createMockStore({})}>
        <NotificationList<string> {...props()} />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
