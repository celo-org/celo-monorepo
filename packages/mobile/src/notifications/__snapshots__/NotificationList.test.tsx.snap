// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationList renders correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "margin": 16,
          }
        }
      >
        <Text>
          test-a
        </Text>
      </View>
    </View>
  </RCTScrollView>
</SafeAreaView>
`;
